### PR TITLE
Add support for appending a unique string to the filename for the wp/v2/media endpoint

### LINKED ISSUE #41643: Add support for appending a unique string to the filename for the `wp/v2/media` endpoint
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

WooCommerce has a setting to `Append a unique string to filename for security`. This seems to work fine when using the uploader available through the WP media modal, but this doesn't work when uploading a file using the `wp/v2/media` endpoint, which Gutenbergs `uploadMedia` [function uses](https://github.com/WordPress/gutenberg/blob/53a44bedfc90451592c895e245bc662f3b345333/packages/media-utils/src/utils/upload-media.js#L218).

We should add this support to the `wp/v2/media` endpoint as well.

Some background:
In order for this to work in the WP media modal we have to send the `type` param with `downloadable_product`.
The ID is being generated as part of this function and hook (`update_filename`): https://github.com/woocommerce/woocommerce/blob/2d44b1f94b7c886093b86cd8e602efde1a13fef5/plugins/woocommerce/includes/admin/class-wc-admin-post-types.php#L783 

Files uploaded with the `downloadable_product` type are also added in the `woocommerce_uploads` folder.

Here is a stack trace of the current work upload process:
```
/app/public/wp-includes/class-wp-hook.php(326): WC_Admin_Post_Types->update_filename('image-6.png', '.png', '/Users/...')
/app/public/wp-includes/plugin.php(205): WP_Hook->apply_filters('image-6.png', Array)
/app/public/wp-includes/functions.php(2790): apply_filters('wp_unique_filen...', 'image-6.png', '.png', '/Users/...', NULL, Array, '')
/app/public/wp-admin/includes/file.php(984): wp_unique_filename('/Users/...', 'image-6.png', NULL)
/app/public/wp-admin/includes/file.php(1105): _wp_handle_upload(Array, Array, '2023-11-21 14:2...', 'wp_handle_uploa...')
/app/public/wp-admin/includes/media.php(306): wp_handle_upload(Array, Array, '2023-11-21 14:2...')
/app/public/wp-admin/includes/ajax-actions.php(2623): media_handle_upload('async-upload', NULL, Array)
/app/public/wp-admin/async-upload.php(33): wp_ajax_upload_attachment()
```

**Acceptance criteria**

- [ ] `wp/v2/media` endpoint has support for the `type` parameter
- [ ] When `type` is `downloadable_product` we append a unique string if the setting is enabled under **WooCommerce > Settings > Products > Downloadable products > Filename**
- [ ] When `type` is `downloadable_product` we add the file within the `woocommerce_uploads` directory within `uploads` ( ex: `/wp-content/uploads/woocommerce_uploads/2023/11/test.png` )
- [ ] Add all the features removed in https://github.com/woocommerce/woocommerce/pull/41621

### PR DESCRIPTION
**Pull request by nathanss**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/41643 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the new product editor at /wp-admin/admin.php?page=wc-settings&tab=advanced&section=features
2. Go to /wp-admin/admin.php?page=wc-settings&tab=products&section=downloadable
3. Check the option "Append a unique string to filename for security" and save.
4. Go to Products > Add 
5. Go to the Downloads section
6. Drop a supported file for upload
7. Check that a unique string was appended to the filename.
8. Click on "Add new" > "Upload"
9. Select a supported file
10. Check that a unique string was appended to the filename.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/API/Init.php [hunk 1]
```diff
@@ -49,6 +49,9 @@ public function __construct() {
 
 		// Add currency symbol to orders endpoint response.
 		add_filter( 'woocommerce_rest_prepare_shop_order_object', array( __CLASS__, 'add_currency_symbol_to_order_response' ) );
+
+		include_once __DIR__ . '/../../../includes/admin/class-wc-admin-upload-downloadable-product.php';
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/API/Init.php]: Would we potentially want to use `WC_ABSPATH . 'includes....` here? instead of the `/../../..`?

**Review by louwie17 — CHANGES_REQUESTED** [review]
Awesome work here @nathanss and nice debugging!!
I assume the include once in the API/Init.php was the ultimate thing that fixed it?

I just left one minor suggestion, otherwise looks good and tested great!!

File: plugins/woocommerce/src/Admin/API/Init.php (hunk 1 above)
Comment (nathanss) [reply plugins/woocommerce/src/Admin/API/Init.php]: Sure, I get lost in all those constants :)

**Comment by nathanss** [comment]
@louwie17 

> I assume the inclusion once in the API/Init.php was the ultimate thing that fixed it?

Yes, it then attaches the already existing hooks that handled this logic for media library uploads :)

**Review by louwie17 — APPROVED** [review]
LGTM 🚀

**Pull request merged**